### PR TITLE
[4.0] Redo the image editing api

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request is the second part for  the Tabs reworked code: https://github.com/joomla/joomla-cms/pull/34813

### Summary of Changes
Some changes to the API:
- Now it's async (used to be synchronous, but that won't work for many cases, current libraries, etc)
- Fixing all the problems with:
- - Empty fields on load
- - Save needed a click away to update the image
- - other inconsistencies

#### This PR is not testable with the downladable version

Apply this PR!!!
You need to run `npm install`

Try to edit some images and using all the existing options

### Actual result BEFORE applying this Pull Request

Many bugs

### Expected result AFTER applying this Pull Request

No (obvious) bugs

### Documentation Changes Required

The old registration was something like:
```js
  // Register the Events
  Joomla.MediaManager.Edit.crop = {
    Activate(mediaData) {
      // Initialize
      initCrop(mediaData);
    },
    Deactivate() {
      if (!Joomla.MediaManager.Edit.crop.cropper) {
        return;
      }
      // Destroy the instance
      Joomla.MediaManager.Edit.crop.cropper.destroy();
    },
  };
```

Which is changed to:

```js
// Register the Events
window.addEventListener('media-manager-edit-init', () => {
  Joomla.MediaManager.Edit.plugins.crop = {
    Activate(image) {
      return new Promise((resolve /* , reject */) => {
        init(image);
        Joomla.MediaManager.Edit.plugins.crop.cropper
          .setAspectRatio(formElements.cropAspectRatioOption.value);
        resolve();
      });
    },
    Deactivate(image) {
      return new Promise((resolve /* , reject */) => {
        if (image.cropper) {
          image.cropper.destroy();
          delete Joomla.MediaManager.Edit.plugins.crop.cropper;
        }
        resolve();
      });
    },
  };
}, { once: true });

```

So most of the changes are just changes in the initialization step:
- there's an event for the registration, on the window, named 'media-manager-edit-init'
- The plugin name is now one level deeper, from `Joomla.MediaManager.Edit.crop` to `Joomla.MediaManager.Edit.plugins.crop`
- Both the `Activate` and `Deactivate` functions should return a promise.

@wilsonge sorry...

### CODE REVIEW

**Comment by dgrammatiko** [comment]
This PR should be fixing:
https://github.com/joomla/joomla-cms/issues/33843
https://github.com/joomla/joomla-cms/issues/31859
https://github.com/joomla/joomla-cms/pull/33192

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on dc65bdbb72158788e34b5a1f367c11d7a7941b39<br /><br />THANK YOU!!!!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34885">issues.joomla.org/tracker/joomla-cms/34885</a>.</sub>

**Comment by Quy** [comment]
Edit an image:
> Uncaught TypeError: this.Undo is undefined edit-images.js:99:5

Click Close button:
> Uncaught TypeError: newForm.task is undefined core.js:389:7

**Comment by dgrammatiko** [comment]
> Uncaught TypeError: this.Undo is undefined edit-images.js:99:5

Fixing the lint issues without checking the code is not a great idea, should be ok now

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on 627b449ec236b38486bb2e0310260e2d788d547e<br /><br />It all seems to work. I did notice some anomalies:
The aspect ratio box default was 'None'. The Edit screen does not have a Help button (it could use the media screen Help). Changing X-axis and Width did not show in the grid. The image is barely visible in 768x1024 screen mode and the grid shrinks strangely. Cannot edit (etc) from list view. In edit mode repeated clicking of Save moves the images left a little with each click (but the saved image is OK).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34885">issues.joomla.org/tracker/joomla-cms/34885</a>.</sub>

**Comment by dgrammatiko** [comment]
> The aspect ratio box default was 'None'. 

Oops I change that in the other PR, will undo the changes here

> The Edit screen does not have a Help button (it could use the media screen Help). 

This is out of the scope of this PR but should be easy to add a help button here: https://github.com/joomla/joomla-cms/blob/193408176112c05055d9e005c0b82070cf0c5aef/administrator/components/com_media/src/View/File/HtmlView.php#L75

> Changing X-axis and Width did not show in the grid.
> The image is barely visible in 768x1024 screen mode and the grid shrinks strangely.
> In edit mode repeated clicking of Save moves the images left a little with each click (but the saved image is OK).

I guess this is for crop, investigating all three issues


> Cannot edit (etc) from list view.

This is something for the Media manager

**Comment by Quy** [comment]
For crop, the image expands height/width accordingly, but not under the other actions. Is this intentional?

**Comment by dgrammatiko** [comment]
> For crop, the image expands height/width accordingly, but not under the other actions. Is this intentional?

The image has a `max-width: 100%` so it should adapt to the parent's size. I didn't touch that part (or the logic), just updated the initialization logic and fixed some obvious mistakes (re attaching listeners, etc). Anyways I will check this again

**Comment by dgrammatiko** [comment]
> For crop, the image expands height/width accordingly, but not under the other actions. Is this intentional?

Should be fine now

**Comment by Quy** [comment]
It is no longer side by side. The settings are at the top and image at the bottom.

**Comment by dgrammatiko** [comment]
> It is no longer side by side. The settings are at the top and image at the bottom.

Oops, sorry, handling too many PRs is not easy

**Comment by Quy** [comment]
The image still expands as seen between crop and resize.

34885-crop

34885-resize

**Comment by dgrammatiko** [comment]
Image size corrected

**Comment by richard67** [comment]
@dgrammatiko It needs to solve conflicts and update the branch here after recent merge into 4.0-dev.

**Comment by dgrammatiko** [comment]
> @dgrammatiko It needs to solve conflicts and update the branch here after recent merge into 4.0-dev.

Yup, and now I hope tests are green

**Comment by richard67** [comment]
@dgrammatiko You might have to update your branch again because just right now George merged the changed tests into 4.0-dev.

**Comment by richard67** [comment]
@dgrammatiko System tests failed so it seems you have updated your branch just 2 minutes too early.

File: build/media_source/com_media/scss/components/_media-edit.scss [hunk 1]
```diff
@@ -8,7 +8,7 @@
   }
 
   joomla-tab-element {
-    grid-template-columns: repeat(4, minmax(240px, 1fr));
+    grid-template-columns: repeat(auto-fit, minmax(240px,1fr));
```
Comment (brianteeman) [inline build/media_source/com_media/scss/components/_media-edit.scss]:
```suggestion
    grid-template-columns: repeat(auto-fit, minmax(240px, 1fr));
```

File: build/media_source/plg_media-action_rotate/js/rotate.es6.js [hunk 2]
```diff
@@ -2,36 +2,31 @@
...
  * @license    GNU General Public License version 2 or later; see LICENSE.txt
  */
-Joomla = window.Joomla || {};
+const activated = false;
 
-Joomla.MediaManager = Joomla.MediaManager || {};
-Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
+// Update image
+const rotate = (angle, image) => {
+  // The canvas where we will resize the image
```
Comment (brianteeman) [inline build/media_source/plg_media-action_rotate/js/rotate.es6.js]:
```suggestion
  // The canvas where we will rotate the image
```

**Comment by richard67** [comment]
All CI checks green ... there must be something wrong ;-)

**Comment by wilsonge** [comment]
I don't think the tests cover much of the image editing :P Mainly the list view :D

**Comment by dgrammatiko** [comment]
After https://github.com/joomla/joomla-cms/pull/34885/commits/1c8e84be00e194d61ceaea369b3a85ae3726557f I think all should be fine

**Comment by richard67** [comment]
@dgrammatiko Javascript CS: https://ci.joomla.org/joomla/joomla-cms/46162/1/21

**Comment by dgrammatiko** [comment]
This should be ready for testing now

**Comment by dgrammatiko** [comment]
> I don't think the tests cover much of the image editing :P Mainly the list view :D

The best tests, navigate to the page read the title. Test is passing. Yeah...

**Comment by wilsonge** [comment]
> The best tests, navigate to the page read the title. Test is passing. Yeah...

Lots of the tests still fail on bad changes though :P What does that say :D

**Comment by dgrammatiko** [comment]
> What does that say :D

Unstable server?

**Comment by wilsonge** [comment]
So the crop filters don't apply - but also don't error if you pick e.g. a portrait crop for a landscape image (so just a general lack of feedback - not sure whether that was the case before or not - hard to tell now the other tabs PR is merged as there's no image preview in 4.0-dev right now.

I also got a 403 back when trying to save it. It's a put request to `http://localhost/~george/joomla-cms/administrator/index.php?option=com_media&format=json&mediatypes=0,1,2,3&task=api.files&path=local-images:/banner-og-github.png` - will test properly later just in case it's something locally. But somehow doubt it as it's just a simple file upload then edit.

**Comment by dgrammatiko** [comment]
@wilsonge make sure you have cleared the browser cache, but this is also a bug

**Comment by Quy** [comment]
Open an image.
Click Resize tab.
Click Save button.
Image saved.
Click Save button.
Error in console: 
> Illegal mime type detected: application/octet-stream

**Comment by wilsonge** [comment]
Screenshot 2021-07-27 at 18 21 14

Hit the 90 degree rotate button twice. You get kinda weird behaviour.

**Comment by dgrammatiko** [comment]
> Hit the 90 degree rotate button twice. You get kinda weird behaviour.

Fixed

**Comment by wilsonge** [comment]
If you manually enter text for the crop it seems to have no affect. If you adjust the crop box on the image it works fine. 

The clicking of 90 degree's rotation is better but not fixed too.

**Comment by dgrammatiko** [comment]
> The clicking of 90 degree's rotation is better but not fixed too.

What's the problem now? The image should cover all the right area (assuming it's big enough)

> If you manually enter text for the crop it seems to have no effect. If you adjust the crop box on the image it works fine.

@Quy also mentioned this but the initialization code for cropper is the same AFAICT

**Comment by wilsonge** [comment]
> What's the problem now? The image should cover all the right area (assuming it's big enough)

Same issue. I just have to click 90 degree rotation 4 times instead of twice :) I'm not speed clicking either to try and break things. Sometimes if you wait long enough the screen area resizes itself back to the correct size.

**Comment by wilsonge** [comment]
I also think the aspect ratio dropdown is broken? Not sure if that's related to the cropper code though

**Comment by dgrammatiko** [comment]
@wilsonge  Crop should be fine again with https://github.com/joomla/joomla-cms/pull/34885/commits/910ac38d0f59f0ba92bcb5ce65689fff75739879

**Comment by dgrammatiko** [comment]
> Same issue. I just have to click 90 degree rotation 4 times instead of twice :) I'm not speed clicking either to try and break things. Sometimes if you wait long enough the screen area resizes itself back to the correct size.

Ok, so this was a 2 part solution:
- there was some weird css on the parent element of the image
- browser needed some time to rethink the layout/painting

Give it some hard reloading before testing

**Comment by wilsonge** [comment]
OK So cropper works great now! Way better. But now from that commit rotation doesn't work. It never leaves the cropper view

**Comment by wilsonge** [comment]
Screenshot 2021-07-27 at 20 34 08

It's all broken 😆

File: build/media_source/com_media/js/edit-images.es6.js [hunk 3]
```diff
@@ -2,304 +2,372 @@
...
+    this.imagePreview.id = 'image-preview';
+    this.imagePreview.style.width = '100%';
+    this.imagePreview.style.height = 'auto';
+    this.imagePreview.style.maxWidth = '100%';
+    this.baseContainer.appendChild(this.imagePreview);
+  }
+
+  async activate(name) {
+    this.imagePreview = null;
+    this.createImageContainer(this.current.contents);
```
Comment (wilsonge) [inline build/media_source/com_media/js/edit-images.es6.js]:
```suggestion
    this.createImageContainer(this.current);
```

So this fixes the initialisation error. But then i get 4 copies of the image loaded at full screen. So it's not really a good fix :)

**Comment by wilsonge** [comment]
Screenshot 2021-07-27 at 22 29 55

I mean rotate may well be acting the same as before. But not sure this is intended :D However it's way better than it has been all evening.

**Comment by wilsonge** [comment]
Thanks!

**Comment by dgrammatiko** [comment]
Release the Kraken

source

**Comment by richard67** [comment]
@dgrammatiko There is still much to be fixed.

**Comment by dgrammatiko** [comment]
@richard67 I know, but a new RC is always a good thing... 😉

**Comment by laoneo** [comment]
Sorry guys, but since this pr the reset button is not working anymore.

**Comment by richard67** [comment]
Has it ever worked? ;-)

**Comment by richard67** [comment]
Sorry, was joke. Could you open an issue?

**Pull request merged**